### PR TITLE
core: replace raw intrinsics with universal intrinsics in copy.cpp

### PR DESCRIPTION
**Pull request by seiko2plus**
### This pullrequest changes
- replace raw intrinsics with universal intrinsics in copy.cpp

````
force_builders=Custom
docker_image:Custom=powerpc64le
````
 #### Merge with extra: opencv/opencv_extra#437

### CODE REVIEW

**Comment by savuor** [comment]
To pass the tests the branch in opencv_extra should have the same name as in opencv repo

**Comment by seiko2plus** [comment]
@savuor both branch have the same name now, sorry :)

File: modules/core/src/copy.cpp [hunk 1]
```diff
@@ -130,31 +129,33 @@ copyMask_<ushort>(const uchar* _src, size_t sstep, const uchar* mask, size_t mst
...
-            uint16x8_t v_src = vld1q_u16(src + x), v_dst = vld1q_u16(dst + x);
-            vst1q_u16(dst + x, vbslq_u16(v_mask_res, v_src, v_dst));
+                v_uint16x8 v_src1 = v_load(src + x), v_src2 = v_load(src + x + 8),
+                           v_dst1 = v_load(dst + x), v_dst2 = v_load(dst + x + 8);
+
+                v_uint8x16 v_nmask1, v_nmask2;
+                v_uint8x16 v_nmask = v_load(mask + x) == v_zero;
+                v_zip(v_nmask, v_nmask, v_nmask1, v_nmask2);
+
+                #if CV_SSE4_2
```
Comment (vpisarev) [inline modules/core/src/copy.cpp]: unfortunately, currently we are not testing this branch; can we remove it for now?

File: modules/core/src/copy.cpp [hunk 2]
```diff
@@ -90,28 +90,27 @@ copyMask_<uchar>(const uchar* _src, size_t sstep, const uchar* mask, size_t mste
...
+        ) {
+            v_uint8x16 v_zero = v_setzero_u8();
+
+            for( ; x <= size.width - 16; x += 16 )
+            {
+                v_uint8x16 v_src   = v_load(src  + x),
+                           v_dst   = v_load(dst  + x),
+                           v_nmask = v_load(mask + x) == v_zero;
+
+                #if CV_SSE4_2
```
Comment (vpisarev) [inline modules/core/src/copy.cpp]: unfortunately, we are not testing this branch currently; can we remove it for now?

**Comment by seiko2plus** [comment]
**os**: Debian GNU/Linux 9 / ppc64le
**compiler**: gcc version 7.3.0

````
processor	: 0
cpu		: POWER8E (raw), altivec supported
clock		: 3425.000000MHz
revision	: 2.1 (pvr 004b 0201)
timebase	: 512000000
platform	: pSeries
model		: IBM pSeries (emulated by qemu)
machine		: CHRP IBM pSeries (emulated by qemu)
MMU		: Hash
````
----------------------------------------------------------------

|Name of Test|before|after|
|---|:-:|:-:|
|Mat_CopyToWithMask::Size_MatType::(127x61, 8UC1)|0.007 ms|0.001 ms|
|Mat_CopyToWithMask::Size_MatType::(127x61, 8UC2)|0.007 ms|0.001 ms|
|Mat_CopyToWithMask::Size_MatType::(1280x720, 8UC1)|0.798 ms|0.063 ms|
|Mat_CopyToWithMask::Size_MatType::(1280x720, 8UC2)|0.802 ms|0.136 ms|
|Mat_CopyToWithMask::Size_MatType::(1920x1080, 8UC1)|1.798 ms|0.149 ms|
|Mat_CopyToWithMask::Size_MatType::(1920x1080, 8UC2)|1.811 ms|0.399 ms|
|Mat_CopyToWithMask::Size_MatType::(640x480, 8UC1)|0.266 ms|0.022 ms|
|Mat_CopyToWithMask::Size_MatType::(640x480, 8UC2)|0.267 ms|0.044 ms|

|Name of Test|before|after|
|---|:-:|:-:|
|Mat_SetToWithMask::Size_MatType::(127x61, 8UC1)|0.012 ms|0.006 ms|
|Mat_SetToWithMask::Size_MatType::(127x61, 8UC2)|0.010 ms|0.004 ms|
|Mat_SetToWithMask::Size_MatType::(1280x720, 8UC1)|0.819 ms|0.080 ms|
|Mat_SetToWithMask::Size_MatType::(1280x720, 8UC2)|0.835 ms|0.162 ms|
|Mat_SetToWithMask::Size_MatType::(1920x1080, 8UC1)|1.836 ms|0.175 ms|
|Mat_SetToWithMask::Size_MatType::(1920x1080, 8UC2)|1.875 ms|0.378 ms|
|Mat_SetToWithMask::Size_MatType::(640x480, 8UC1)|0.277 ms|0.031 ms|
|Mat_SetToWithMask::Size_MatType::(640x480, 8UC2)|0.281 ms|0.057 ms|

**Comment by seiko2plus** [comment]
**os**: Debian GNU/Linux 9 / x86_64
**Compiler**: gcc version 6.3.0

````
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 60
model name	: Intel Core Processor (Haswell, no TSX)
stepping	: 1
microcode	: 0x1
cpu MHz		: 2394.472
cache size	: 4096 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl eagerfpu pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm invpcid_single kaiser tpr_shadow vnmi flexpriority ept vpid fsgsbase bmi1 avx2 smep bmi2 erms invpcid xsaveopt
bugs		:
bogomips	: 4788.94
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:
````
----------------------------------------------------------------

#### without IPP

|Name of Test|before|after|
|---|:-:|:-:|
|Mat_CopyToWithMask::Size_MatType::(127x61, 8UC1)|0.006 ms|0.001 ms|
|Mat_CopyToWithMask::Size_MatType::(127x61, 8UC2)|0.007 ms|0.001 ms|
|Mat_CopyToWithMask::Size_MatType::(1280x720, 8UC1)|0.782 ms|0.110 ms|
|Mat_CopyToWithMask::Size_MatType::(1280x720, 8UC2)|0.806 ms|0.205 ms|
|Mat_CopyToWithMask::Size_MatType::(1920x1080, 8UC1)|1.834 ms|0.239 ms|
|Mat_CopyToWithMask::Size_MatType::(1920x1080, 8UC2)|1.970 ms|0.467 ms|
|Mat_CopyToWithMask::Size_MatType::(640x480, 8UC1)|0.260 ms|0.036 ms|
|Mat_CopyToWithMask::Size_MatType::(640x480, 8UC2)|0.255 ms|0.062 ms|

|Name of Test|before|after|
|---|:-:|:-:|
|Mat_SetToWithMask::Size_MatType::(127x61, 8UC1)|0.009 ms|0.003 ms|
|Mat_SetToWithMask::Size_MatType::(127x61, 8UC2)|0.008 ms|0.003 ms|
|Mat_SetToWithMask::Size_MatType::(1280x720, 8UC1)|0.778 ms|0.105 ms|
|Mat_SetToWithMask::Size_MatType::(1280x720, 8UC2)|0.801 ms|0.171 ms|
|Mat_SetToWithMask::Size_MatType::(1920x1080, 8UC1)|1.832 ms|0.244 ms|
|Mat_SetToWithMask::Size_MatType::(1920x1080, 8UC2)|1.845 ms|0.415 ms|
|Mat_SetToWithMask::Size_MatType::(640x480, 8UC1)|0.258 ms|0.036 ms|
|Mat_SetToWithMask::Size_MatType::(640x480, 8UC2)|0.262 ms|0.056 ms|

-------------------------------------------------------------------------------------

#### without IPP, with SSE42

|Name of Test|before|after|
|---|:-:|:-:|
|Mat_CopyToWithMask::Size_MatType::(127x61, 8UC1)|0.001 ms|0.001 ms|
|Mat_CopyToWithMask::Size_MatType::(127x61, 8UC2)|0.002 ms|0.001 ms|
|Mat_CopyToWithMask::Size_MatType::(1280x720, 8UC1)|0.106 ms|0.107 ms|
|Mat_CopyToWithMask::Size_MatType::(1280x720, 8UC2)|0.211 ms|0.187 ms|
|Mat_CopyToWithMask::Size_MatType::(1920x1080, 8UC1)|0.240 ms|0.235 ms|
|Mat_CopyToWithMask::Size_MatType::(1920x1080, 8UC2)|0.476 ms|0.440 ms|
|Mat_CopyToWithMask::Size_MatType::(640x480, 8UC1)|0.036 ms|0.036 ms|
|Mat_CopyToWithMask::Size_MatType::(640x480, 8UC2)|0.067 ms|0.062 ms|

|Name of Test|before|after|
|---|:-:|:-:|
|Mat_SetToWithMask::Size_MatType::(127x61, 8UC1)|0.003 ms|0.003 ms|
|Mat_SetToWithMask::Size_MatType::(127x61, 8UC2)|0.003 ms|0.003 ms|
|Mat_SetToWithMask::Size_MatType::(1280x720, 8UC1)|0.100 ms|0.100 ms|
|Mat_SetToWithMask::Size_MatType::(1280x720, 8UC2)|0.211 ms|0.161 ms|
|Mat_SetToWithMask::Size_MatType::(1920x1080, 8UC1)|0.218 ms|0.217 ms|
|Mat_SetToWithMask::Size_MatType::(1920x1080, 8UC2)|0.498 ms|0.374 ms|
|Mat_SetToWithMask::Size_MatType::(640x480, 8UC1)|0.034 ms|0.034 ms|
|Mat_SetToWithMask::Size_MatType::(640x480, 8UC2)|0.070 ms|0.054 ms|

**Comment by seiko2plus** [comment]
**os**: Debian GNU/Linux 9 / armv8l
**compiler**: gcc version 6.3.0
**cpu**: Exynos Octa 7870 1.59GHZ

````
processor	: 0
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt lpae evtstrm aes pmull sha1 sha2 crc32
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4

````

|Name of Test|before|after|
|---|:-:|:-:|
|Mat_CopyToWithMask::Size_MatType::(127x61, 8UC1)|0.006 ms|0.006 ms|
|Mat_CopyToWithMask::Size_MatType::(127x61, 8UC2)|0.013 ms|0.010 ms|
|Mat_CopyToWithMask::Size_MatType::(1280x720, 8UC1)|1.805 ms|1.953 ms|
|Mat_CopyToWithMask::Size_MatType::(1280x720, 8UC2)|2.970 ms|2.704 ms|
|Mat_CopyToWithMask::Size_MatType::(1920x1080, 8UC1)|4.945 ms|3.968 ms|
|Mat_CopyToWithMask::Size_MatType::(1920x1080, 8UC2)|7.289 ms|7.339 ms|
|Mat_CopyToWithMask::Size_MatType::(640x480, 8UC1)|0.497 ms|0.503 ms|
|Mat_CopyToWithMask::Size_MatType::(640x480, 8UC2)|0.995 ms|0.903 ms|


|Name of Test|before|after|
|---|:-:|:-:|
|Mat_SetToWithMask::Size_MatType::(127x61, 8UC1)|0.009 ms|0.009 ms|
|Mat_SetToWithMask::Size_MatType::(127x61, 8UC2)|0.016 ms|0.013 ms|
|Mat_SetToWithMask::Size_MatType::(1280x720, 8UC1)|1.173 ms|1.174 ms|
|Mat_SetToWithMask::Size_MatType::(1280x720, 8UC2)|2.057 ms|1.798 ms|
|Mat_SetToWithMask::Size_MatType::(1920x1080, 8UC1)|2.810 ms|2.809 ms|
|Mat_SetToWithMask::Size_MatType::(1920x1080, 8UC2)|4.578 ms|4.012 ms|
|Mat_SetToWithMask::Size_MatType::(640x480, 8UC1)|0.267 ms|0.264 ms|
|Mat_SetToWithMask::Size_MatType::(640x480, 8UC2)|0.655 ms|0.538 ms|

File: modules/core/src/copy.cpp (hunk 1 above)
Comment (seiko2plus) [reply modules/core/src/copy.cpp]: I did some perf tests. if it isn't enough, sure we can remove it for now.

Comment (alalek) [reply modules/core/src/copy.cpp]: I believe it is better to move this branch into `v_select()` implementation directly. See [this example](https://github.com/opencv/opencv/blob/3.4.0/modules/core/include/opencv2/core/hal/intrin_sse.hpp#L1179-L1191).

Comment (seiko2plus) [reply modules/core/src/copy.cpp]: I don't think so. v_select() is a bitwise select but mask in _mm_blendv_epi8() must be packed as 8-bit of 0x00 or 0xFF.
Should I add new universal intrinsic? or just remove this branch?
Also I made a mistake here _mm_blendv_epi8() is part of SSE4.1 not SSE4.2.

**Comment by vpisarev** [comment]
great! :+1:

File: modules/core/src/copy.cpp (hunk 1 above)
Comment (vpisarev) [reply modules/core/src/copy.cpp]: for now using `_mm_blendv_epi8()` is fine. Later on we might construct the proper universal intrinsic with the same efficiency

**Pull request merged**